### PR TITLE
[rush] Add support for "Current" Node version (12.x)

### LINKED ISSUE #1373: [rush] Add support for "Current" Node version (12.x)
Please add support for the "Current" Node version (`12.x`), in addition to the "Active LTS" (`10.x`) and "Maintenance LTS" versions (`8.x`).

https://github.com/microsoft/web-build-tools/blob/b31c2e73ce549ce37531ef39034c860cf5782cd4/apps/rush/src/start.ts#L22-L36

### PR DESCRIPTION
**Pull request by mikeharder**
- Remove version warnings for Node 12
- Do warn for Node 6 which is End-of-Life
- Add Node 12 to test matrix
- Fixes #1373

### CODE REVIEW

File: common/config/azure-pipelines/ci.yaml [hunk 1]
```diff
@@ -2,12 +2,13 @@ jobs:
 - job: PRBuild
   condition: succeeded()
   strategy:
-    maxParallel: 2
```
Comment (mikeharder) [inline common/config/azure-pipelines/ci.yaml]: If `maxParallel` is not specified, the default behavior is to run all matrix rows in parallel.

Comment (sachinjoseph) [inline common/config/azure-pipelines/ci.yaml]: Is removing this line intentional?

**Comment by mikeharder** [comment]
I have a few more changes incoming:

- Update CHANGELOG
- Add Node 12 to list of versions in root `rush.json`

File: core-build/gulp-core-build-sass/package.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "autoprefixer": "~9.1.3",
     "postcss-modules": "~1.3.1",
     "glob": "~7.0.5",
-    "node-sass": "4.9.3",
+    "node-sass": "4.12.0",
```
Comment (mikeharder) [inline core-build/gulp-core-build-sass/package.json]: `node-sass@4.12.0` is required for Node 12 support:

https://github.com/sass/node-sass/releases/tag/v4.12.0

File: core-build/gulp-core-build-sass/package.json [hunk 3]
```diff
@@ -32,7 +32,7 @@
     "@microsoft/rush-stack-compiler-3.2": "0.3.24",
     "@microsoft/node-library-build": "6.0.72",
     "@types/glob": "5.0.30",
-    "@types/node-sass": "3.10.32",
+    "@types/node-sass": "4.11.0",
```
Comment (mikeharder) [inline core-build/gulp-core-build-sass/package.json]: Updating to latest since I am also updating `node-sass` to latest

**Comment by mikeharder** [comment]
Next issue is that Gulp 3 is not compatible with Node 12.  We will need to update to Gulp 4.

**Comment by mikeharder** [comment]
@iclanton, @octogonz, @sachinjoseph: By my quick glance, this repo appears to have a strong dependency on `gulp`, so updating from `gulp@3.9.1` to `gulp@4.0.2` (which is required to support Node 12) may be non-trivial.  I believe this work will need to be done at some point between now and Node 12 LTS (planned for 2019-10-22), but this PR may not be the right place.

Since this PR adds Node 12 to the test matrix, it requires the whole monorepo to be compatible with Node 12, not just Rush.

I see a few options:

1. Update the Rush version checks to allow Node 12, but don't add Node 12 to the test matrix.  This prevents users from seeing the warning on Node 12, but Rush would not actually be testing using Node 12.
2. When testing Node 12, only build Rush rather than the whole repo.  We may need to update the Rush dev dependency to `gulp@4.0.2`, but the other projects in the repo could remain on `gulp@3.9.1`.
3. Update the whole repo to `gulp@4.0.2`.

I will attempt to update the whole repo to `gulp@4.0.2` and see what happens.

**Comment by mikeharder** [comment]
I am only seeing one error after updating `gulp`:

```
FAILURE (1)                                                                                                                                           
================================                                                                                                                      
web-library-build-test (7.45 seconds)                                                                                                                 
[17:16:38] 'default' errored after 114 ms                                                                                                             
[17:16:38] Error: File not found with singular glob: D:/Git/web-build-tools/build-tests/web-library-build-test/preCopyTest.js (if this was purposeful,
 use `allowEmpty` option)                                                                                                                             
    at Glob.<anonymous> (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob-stream\6.1.0\node_modules\glob-stream\readable.js:84
:17)                                                                                                                                                  
    at Object.onceWrapper (events.js:291:20)                                                                                                          
    at Glob.emit (events.js:203:13)                                                                                                                   
    at Glob.EventEmitter.emit (domain.js:471:20)                                                                                                      
    at Glob._finish (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:197:8)                  
    at done (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:182:14)                         
    at Glob._processSimple2 (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:688:12)         
    at D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:676:10                                
    at Glob._stat2 (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:772:12)                  
    at lstatcb_ (D:\Git\web-build-tools\common\temp\node_modules\.registry.npmjs.org\glob\7.1.4\node_modules\glob\glob.js:764:12)                     
Process terminated before summary could be written, possible error in async code not continuing!                                                      
================================                                                                                                                      
```

I think this might be a known change in Gulp 4:

https://github.com/gulpjs/vinyl-fs/issues/292

File: core-build/gulp-core-build-sass/package.json (hunk 2 above)
Comment (octogonz) [reply core-build/gulp-core-build-sass/package.json]: Does this introduce a `node-gyp` dependency?  See recent discussion in https://github.com/microsoft/web-build-tools/issues/1361

**Comment by iclanton** [comment]
Thanks for doing all of this work! I'll take a look at this branch tonight and see if I can address the remaining issues.

**Comment by mikeharder** [comment]
@iclanton: You're welcome.  However, note that I'm not very familiar with any of the dependencies I updated.  I just know the previous versions are not compatible with Node 12, so I attempted to update to latest.  I don't know what impact this might have on the consuming packages, or whether this is the overall direction the repository should be going.

File: core-build/gulp-core-build-sass/package.json (hunk 2 above)
Comment (iclanton) [reply core-build/gulp-core-build-sass/package.json]: Looks like we're good on that front. [`node-sass@4.9.3`](https://github.com/sass/node-sass/releases/tag/v4.9.3) doesn't support Node 12 natively:

image

But [`node-sass@4.12.0`](https://github.com/sass/node-sass/releases/tag/v4.12.0) does:

image

**Comment by iclanton** [comment]
Before we merge this in, I want to test it against SPFx. If that looks good, this should be good to go.

**Comment by octogonz** [comment]
@mikeharder  BTW we aren't expecting you to make all the changes suggested in https://github.com/microsoft/web-build-tools/issues/1373.  Ian volunteered to help out with that part.

**Comment by mikeharder** [comment]
@octogonz, @iclanton: I would suggest splitting this work into multiple PRs:

1. One PR for the Rush changes (warning messages, setting to suppress messages, etc)
2. Separate PRs to update dependencies and other changes required to make `web-build-tools` as a whole compatible with Node 12
3. Enable Node 12 in test matrix

**Comment by iclanton** [comment]
I'm down with that. Here are PRs:

1. https://github.com/microsoft/web-build-tools/pull/1388
2. https://github.com/microsoft/web-build-tools/pull/1387
3. https://github.com/microsoft/web-build-tools/pull/1386

**Comment by mikeharder** [comment]
Replaced by #1386, #1387, and #1388

**Pull request closed without merging**